services/data_sources/json_notification_data_source.rb

```diff
@@ -0,0 +1,162 @@
+require "securerandom"
+require_relative "notification_data_source"
+require_relative "../../shared/logging_module"
+require_relative "../../shared/json_convertible"
+require_relative "../../shared/models/notification"
+
+module FastlaneCI
+  # Mixin the JSONConvertible class for Notification
+  class Notification
+    include FastlaneCI::JSONConvertible
+  end
+
+  # Data source for notifications backed by JSON
+  class JSONNotificationDataSource < NotificationDataSource
+    include FastlaneCI::JSONDataSource
+    include FastlaneCI::Logging
+
+    class << self
+      attr_accessor :file_semaphore
+    end
+
+    # Can't have us reading and writing to a file at the same time
+    JSONNotificationDataSource.file_semaphore = Mutex.new
+
+    # Reloads notifications from the notifications data source after instantiation
+    #
+    # @param  [Any] params
+    # @return [nil]
+    def after_creation(**params)
+      if params.nil?
+        raise "Either user or a provider credential is mandatory."
+      else
+        if !params[:user] && !params[:provider_credential]
+          raise "Either user or a provider credential is mandatory."
+        else
+          params[:provider_credential] ||= params[:user].provider_credential(type: ProviderCredential::PROVIDER_CREDENTIAL_TYPES[:github])
+          @git_repo = FastlaneCI::GitRepo.new(git_config: self.json_folder_path, provider_credential: params[:provider_credential])
+        end
+      end
+
+      reload_notifications
+    end
+
+    # Returns an array of notifications from the notifications JSON file stored
+    # in the configuration git repo
+    #
+    # @return [Array[Notification]]
+    def notifications
+      JSONProjectDataSource.projects_file_semaphore.synchronize do
+        path = git_repo.file_path("notifications.json")
+        return [] unless File.exist?(path)
+
+        return JSON.parse(File.read(path)).map(&Notification.method(:from_json!))
+      end
+    end
+
+    # Writes the notifications array to the git repo as JSON and commit them as
+    # the CI user
+    #
+    # @param  [Array[Notification]] notifications
+    # @return [nil]
+    def notifications=(notifications)
+      JSONNotificationDataSource.file_semaphore.synchronize do
+        File.write(notifications_file_path, JSON.pretty_generate(notifications.map(&:to_object_dictionary)))
+        git_repo.commit_changes!
+      end
+    end
+
+    # Returns `true` if the notification exists in the in-memory notifications object
+    #
+    # @param  [String] name
+    # @return [Boolean]
+    def notification_exist?(name: nil)
+      notification = @notifications.select { |n| n.primary_key == Notification.make_primary_key(name) }.first
```

Comment:
we might need to synchronize on this too, depending on the behavior of `.select`, I know `@notifications` could be updated while this is iterating.